### DISCUSSION TITLE
Deploying projects w/i a monorepo?

### DISCUSSION THREAD

**Discussion by claytonrcarter** [deployphp/deployer#2691] [Help needed]
Aside from manually rewriting a recipe and/or all of it's tasks, are there any recommendations for deploying projects that reside w/i subdirectories of a repository? 

Deploying the code and setting up the symlinks seems to be pretty easy, as is getting the web server to point to the correct location, but once we start using recipes (eg `laravel`, `npm`, etc) things get messy. For instance, the `npm:install` task just does `run("cd {{release_path}} && {{bin/npm}} install");`, but we would need it to `cd {{release_path}}/our-laravel-app`. It's pretty easy to just overwrite this particular task, but working w/ the `laravel` recipe is more complicated: most of the tasks in the Laravel recipe use the [`artisan()` function](https://github.com/deployphp/deployer/blob/master/recipe/laravel.php#L43), which is hard coded to look for either `{{current_path}}/artisan` or `{{release_or_current_path}}/artisan`. Because `artisan` is a plain PHP function, though, it can't just be overwritten because that triggers `PHP Fatal error:  Cannot redeclare Deployer\artisan()`. 

The quick and dirty way around this is to just copy the laravel recipe into each of our projects and alter 2-3 lines of code to point to, eg, `{{release_or_current_path}}/our-laravel-app/artisan`. This is doable, of course, but cumbersome.

Are there any easier ways to do this?

For context, our structure is something like this:

```
/ (repo base)
/laravel-app-1
/laravel-app-2
/shared-package
```

We have been happily using deployer for years w/ these as separate projects in separate repos, but we are looking at consolidating all of our projects into a monorepo to reduce the burden of managing all of these. The general idea is that, instead of having a repo w/ a single Laravel app in it, we would have one repo w/ subdirectories, each w/ a separate Laravel app (or shared package) w/i it.

**Comment by antonmedv** [comment]
There is a feature request to checkout only a subdir #2570. This will be ideal for this case. 
Two options:
- Override deploy:update_code to upload only needed subdir.
- Create a PR with support for this feature. This should be relatively easy.

**Reply by claytonrcarter** [reply]
Thanks Anton, but I think that my use case is slightly different, because I want to keep *all* of the directories in the repo, but only run app-related commands and symlink to one of the dirs.

In the repo dir structure I mentioned above, each of the apps is using the code w/i `/shared-package`, and it's very easy to configure Composer to install the shared package from a local directory (ie `../shared-package`) vs trying to figure out splitting it into  separate private package, etc

Currently, I'm just copying/replacing the recipes that I need (ie `laravel` and `npm`) and updating them to use the following set of config variables:

```
set('project_subdir', 'laravel-app');
set('release_project_path', '{{release_path}}/{{project_subdir}}');
set('current_project_path', '{{current_path}}/{{project_subdir}}');
set('release_or_current_project_path', '{{release_or_current_path}}/{{project_subdir}}');
```

This way I can use `{{release_path}}` to continue referring to the repo as a whole, or `{{release_project_path}}` to refer just to the project or app I'm trying to deploy. (eg for `npm:install`, instead of `run("cd {{release_path}} && {{bin/npm}} install")`, I have it doing `run("cd {{release_project_path}} && {{bin/npm}} install")`.

Another note is that the `cachetool` recipe made this trivial by exposing the `bin/cachetool` setting (so that I could just `set('bin/cachetool', 'landscape/vendor/bin/cachetool')`) though other recipes don't seem to do that reliably.

**Reply by antonmedv** [reply]
Got it, so I think the copy-pasting an existing recipe is the best solution for your case.
